# DS-SF-36 | 04 | Databases and Scrapping | Assignment | Starter Code

## `SQLite` and Bistro

In this assignment, we will be exploring the `bistro` dataset.  The previous assignment used `pandas`.  Today, we'll answer the same questions but using `SQLite`.  In some situations, `pandas` will be a better solution.  In others, doing it using `SQL` will make more sense.  As you gain more experience, you'll know which one to use.

> ### Question 1.  Import the `sqlite3` package.

In [1]:
import os

import pandas as pd
pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

# TODO
import sqlite3

> ### Question 2.  Connect to the `dataset-04-bistro.db` database.  The rest of this assignment focus on the `bistro` table.

In [2]:
# TODO
db = sqlite3.connect(os.path.join('..', 'datasets', 'dataset-04-bistro.db'))

> ### Question 3.  How many samples (i.e., rows) are in this dataset?

In [7]:
# TODO
pd.io.sql.read_sql(
'''
SELECT count()
    FROM bistro
;
''', con = db)

count()
0      244

Answer: 244

> ### Question 4.  Print the first two rows of the table to the console.

In [8]:
# TODO
pd.io.sql.read_sql(
'''
SELECT *
    FROM bistro
    LIMIT 2
;
''', con = db)

index     day    time      name  gender  is_smoker  party  check   tip
0      0  Sunday  Dinner  Kimberly  Female          0      2  16.99  1.01
1      1  Sunday  Dinner  Nicholas    Male          0      3  10.34  1.66

> ### Question 5.  For which week days does the dataset has data for?

In [9]:
# TODO
pd.io.sql.read_sql(
'''
SELECT count(), day
    FROM bistro
    GROUP BY day
;
''', con = db)

count()       day
0       19    Friday
1       87  Saturday
2       76    Sunday
3       62  Thursday

Answer: Dataset has data Thursday through Sunday.

> ### Question 6.  How often was the bistro patronized for each week day?

In [10]:
# TODO
pd.io.sql.read_sql(
'''
SELECT count(), day
    FROM bistro
    GROUP BY day
;
''', con = db)

count()       day
0       19    Friday
1       87  Saturday
2       76    Sunday
3       62  Thursday

Answer: See above

> ### Question 7.  How much tip did waiters collect for each week day?

In [11]:
# TODO
pd.io.sql.read_sql(
'''
SELECT SUM(tip), day
    FROM bistro
    GROUP BY day
;
''', con = db)

SUM(tip)       day
0     51.96    Friday
1    260.40  Saturday
2    247.39    Sunday
3    171.83  Thursday

Answer: See above

> ### Question 8.  What is the average tip per check (in absolute \$) for each week day?

In [12]:
# TODO
pd.io.sql.read_sql(
'''
SELECT AVG(tip), day
    FROM bistro
    GROUP BY day
;
''', con = db)

AVG(tip)       day
0  2.734737    Friday
1  2.993103  Saturday
2  3.255132    Sunday
3  2.771452  Thursday

Answer: TODO

> ### Question 9.  What is the average tip per check (as a percentage of the check) for each week day?

(`CHECK` is a reserved keywork; use `` `check` `` (put the name between backticks) to reference the `check` column)

In [13]:
# TODO
pd.io.sql.read_sql(
'''
SELECT SUM(tip)/SUM([CHECK]), day
    FROM bistro
    GROUP BY day
;
''', con = db)

SUM(tip)/SUM(       day
0       0.159445    Friday
1       0.146424  Saturday
2       0.152038    Sunday
3       0.156732  Thursday

Answer: See above

> ### Question 10.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

In [16]:
# TODO
pd.io.sql.read_sql(
'''
SELECT M.name from
(SELECT DISTINCT name
    FROM bistro
    where gender='Male') as M
    JOIN
(SELECT DISTINCT name
    FROM bistro
    where gender='Female') as F
on m.name=f.name    
;
''', con = db)

name
0  Casey

Answer: casey

> ### Question 11.  If no patrons share the same name, how many unique patrons are in the dataset?

In [34]:
# TODO
pd.io.sql.read_sql(
'''
SELECT SUM(cnt) from(
SELECT COUNT() as cnt, gender
    FROM bistro
    GROUP BY gender)
;
''', con = db)

SUM(cnt)
0       244

Answer: 244

> ### Question 12.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [36]:
# TODO
pd.io.sql.read_sql(
'''
SELECT COUNT(), name
    FROM bistro
    WHERE name in ('Kevin','Alice')
    GROUP BY name
;
''', con = db)

COUNT()   name
0        2  Alice
1        4  Kevin

Answer: Alice twice, Kevin 4 times

> ### Question 13.  Who are the top 3 female and male patrons?

In [53]:
# TODO
pd.io.sql.read_sql(
'''
SELECT * FROM (
SELECT COUNT() as cnt , name
    FROM bistro    
    WHERE gender='Female'
    GROUP BY name
) 
ORDER BY cnt DESC
LIMIT 3

;
''', con = db)

cnt   name
0    4   Mary
1    3  Casey
2    3  Laura

In [54]:
# TODO
pd.io.sql.read_sql(
'''
SELECT * FROM (
SELECT COUNT() as cnt , name
    FROM bistro    
    WHERE gender='Male'
    GROUP BY name
) 
ORDER BY cnt DESC
LIMIT 3

;
''', con = db)

cnt   name
0    8  David
1    5  Casey
2    5  James

Answer: See above

In [55]:
# TODO
pd.io.sql.read_sql(
'''
SELECT * FROM (
SELECT COUNT() as cnt , gender, name
    FROM bistro    
    GROUP BY gender, name
) 
ORDER BY cnt DESC
LIMIT 3

;
''', con = db)

cnt gender   name
0    8   Male  David
1    5   Male  Casey
2    5   Male  James

> ### Question 14.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

In [62]:
# TODO
pd.io.sql.read_sql(
'''
WITH best_tipper as (
SELECT * FROM (
SELECT SUM(tip)/SUM([CHECK]) as tip_perc, name
    FROM bistro    
    GROUP BY name
) 
ORDER BY tip_perc DESC
LIMIT 1
)

SELECT count(*) as cnt, a.name, a.tip_perc
    FROM best_tipper a
    JOIN bistro b
    ON a.name=b.name
;
''', con = db)

cnt     name  tip_perc
0    1  Maryann  0.416667

In [63]:
# TODO
pd.io.sql.read_sql(
'''
WITH worst_tipper as (
SELECT * FROM (
SELECT SUM(tip)/SUM([CHECK]) as tip_perc, name
    FROM bistro    
    GROUP BY name
) 
ORDER BY tip_perc 
LIMIT 1
)

SELECT count(*) as cnt, a.name, a.tip_perc
    FROM worst_tipper a
    JOIN bistro b
    ON a.name=b.name
;
''', con = db)

cnt    name  tip_perc
0    1  Jeremy  0.035638

Answer: TODO